src/graph.rs

```diff
@@ -86,6 +92,131 @@ impl<'tree> Graph<'tree> {
 
         DisplayGraph(self, ctx)
     }
+
+    pub fn display_json<'a>(&'a self, ctx: &'a Context) -> () {
+        // TODO: move this all into a new module
+        struct JSONGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+        struct JSONNode<'a>(usize, &'a GraphNode, &'a Context);
+        struct JSONEdge<'a>(&'a u32, &'a Edge, &'a Context);
+        struct JSONAttributes<'a>(&'a Attributes, &'a Context);
+        struct JSONIdentifier<'a>(&'a Identifier, &'a Context);
+        struct JSONValue<'a>(&'a Value, &'a Context);
+
+        impl<'a, 'tree> ser::Serialize for JSONGraph<'a, 'tree> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let graph = self.0;
+                let ctx = self.1;
+                let mut seq = serializer.serialize_seq(Some(graph.graph_nodes.len()))?;
+                for (node_index, node) in graph.graph_nodes.iter().enumerate() {
+                    seq.serialize_element(&JSONNode(node_index, node, ctx))?;
+                }
+                seq.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONNode<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let node_index = self.0;
+                let node = self.1;
+                let ctx = self.2;
+                // serializing as a map instead of a struct so we don't have to encode a struct name
```

Comment:
It's tempting to use a struct given that the shape of the map in question is static, but serializers would then bloat how the values are encoded with their type name despite their only ever occurring in a homogeneously typed position.

Reply:
That seems the right way, and also more idomatic JSON as a result.

Reply:
I believe serde has a way to "flatten" a struct (or maybe the field of a struct?) to remove that extra level of naming, but I don't suggest blocking this PR on researching and changing to that.